src/EFCore.PG/Query/ExpressionTranslators/Internal/NpgsqlJsonDomTranslator.cs

```diff
@@ -0,0 +1,195 @@
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Reflection;
+using System.Text.Json;
+using Microsoft.EntityFrameworkCore.Query;
+using Microsoft.EntityFrameworkCore.Query.SqlExpressions;
+using Microsoft.EntityFrameworkCore.Storage;
+using Npgsql.EntityFrameworkCore.PostgreSQL.Extensions;
+using Npgsql.EntityFrameworkCore.PostgreSQL.Query.Expressions.Internal;
+using Npgsql.EntityFrameworkCore.PostgreSQL.Query.Internal;
+using Npgsql.EntityFrameworkCore.PostgreSQL.Storage.Internal.Mapping;
+
+namespace Npgsql.EntityFrameworkCore.PostgreSQL.Query.ExpressionTranslators.Internal
+{
+    public class NpgsqlJsonDomTranslator : IMemberTranslator, IMethodCallTranslator
+    {
+        static readonly MemberInfo RootElement = typeof(JsonDocument).GetProperty(nameof(JsonDocument.RootElement));
+        static readonly MethodInfo GetProperty = typeof(JsonElement).GetRuntimeMethod(nameof(JsonElement.GetProperty), new[] { typeof(string) });
+        static readonly MethodInfo GetArrayLength = typeof(JsonElement).GetRuntimeMethod(nameof(JsonElement.GetArrayLength), Type.EmptyTypes);
+
+        static readonly MethodInfo ArrayIndexer = typeof(JsonElement).GetProperties()
+            .Single(p => p.GetIndexParameters().Length == 1 && p.GetIndexParameters()[0].ParameterType == typeof(int))
+            .GetMethod;
+
+        static readonly string[] GetMethods =
+        {
+            nameof(JsonElement.GetBoolean),
+            nameof(JsonElement.GetDateTime),
+            nameof(JsonElement.GetDateTimeOffset),
+            nameof(JsonElement.GetDecimal),
+            nameof(JsonElement.GetDouble),
+            nameof(JsonElement.GetGuid),
+            nameof(JsonElement.GetInt16),
+            nameof(JsonElement.GetInt32),
+            nameof(JsonElement.GetInt64),
+            nameof(JsonElement.GetSingle),
+            nameof(JsonElement.GetString),
+            nameof(JsonElement.GetInt32),
+            nameof(JsonElement.GetInt32),
+            nameof(JsonElement.GetInt32)
```

Comment:
Too many `GetInt32`. Replace them by getters for unsigned integers.

Reply:
Removed. I don't think unsigned integers will work as PostgreSQL doesn't support them (except for some internal types). Leaving these out for now, we can revisit if someone requests support.